portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -190,4 +236,94 @@ private String getUploadedFileUrl(HttpServletRequest request, String formInputEl
+    while (continueReading) {
+      // end index is inclusive, so we have to subtract 1 to get fetchSize bytes
+      byte[] b =
+          blobstoreService.fetchData(blobKey, currentByteIndex, currentByteIndex + fetchSize - 1);
+      outputBytes.write(b);
+
+      // if we read fewer bytes than we requested, then we reached the end
+      if (b.length < fetchSize) {
+        continueReading = false;
+      }
+
+      currentByteIndex += fetchSize;
+    }
+
+    return outputBytes.toByteArray();
+  }
+
+  /**
+   * Uses the Google Cloud Vision API to generate a list of labels that apply to the image
+   * represented by the binary data stored in imgBytes.
+   */
+  private String getImageLabels(byte[] imgBytes) throws IOException {
+    ByteString byteString = ByteString.copyFrom(imgBytes);
+    Image image = Image.newBuilder().setContent(byteString).build();
+
+    Feature feature = Feature.newBuilder().setType(Feature.Type.LABEL_DETECTION).build();
+    AnnotateImageRequest request =
+        AnnotateImageRequest.newBuilder().addFeatures(feature).setImage(image).build();
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    requests.add(request);
+
+    ImageAnnotatorClient client = ImageAnnotatorClient.create();
+    BatchAnnotateImagesResponse batchResponse = client.batchAnnotateImages(requests);
+    client.close();
+    List<AnnotateImageResponse> imageResponses = batchResponse.getResponsesList();
+    AnnotateImageResponse imageResponse = imageResponses.get(0);
+
+    if (imageResponse.hasError()) {
+      System.err.println("Error getting image labels: " + imageResponse.getError().getMessage());
+      return null;
+    }
+
+    return convertToImageLabels(imageResponse.getLabelAnnotationsList());
+  }
+
+  // Converts labels to a json format that is good for storage and sending.
+  private String convertToImageLabels(List<EntityAnnotation> entityLabels) {
+    List<ImageLabel> imageLabels = new ArrayList<>(); 
+    for (EntityAnnotation label : entityLabels) {
+      ImageLabel newLabel = new ImageLabel(label.getDescription(), label.getScore());
+      imageLabels.add(newLabel);
+    }
+    return gson.toJson(imageLabels);
+  }
+
+  private String getDummyImageLabels(byte[] imgBytes) {
+    List<ImageLabel> dummyImageLabels = new ArrayList<>();
+    // Will choose random descriptions from the following array.
+    String[] descriptions = {"Cat", "Dog", "Car", "Scyscraper", "Wagon", 
+                             "Woman", "Man", "Baby", "Octopus", "City", "Sky"};
+    // Choose random number of labels to insert in dummy label array, between
+    // 1 and 5.
+    Random rand = new Random();
+    int randomNum = rand.nextInt(5) + 1;
+    for (int i = 0; i<randomNum; i++) {
+      String randomDescription = descriptions[rand.nextInt(11)];
+      float randomScore = round(rand.nextFloat(), 2);
+      ImageLabel dummyLabel = new ImageLabel(randomDescription, randomScore);
+      dummyImageLabels.add(dummyLabel);
```

Comment:
as above, you can just put the new ImageLabel into the dummyImageLabels.add() line below since it's clear what's happening and you're not using the variable.